### PR TITLE
[onert] Implement ReduceMean train backend

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit implements ReduceMean train backend.

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

Draft: #12358

### CODE REVIEW

File: runtime/onert/backend/train/ops/MeanLayer.cc [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  nnfw::cker::Shape keep_dim_shape;
+  if (_keep_dims == false)
+  {
+    keep_dim_shape.ReplaceWith(getShape(_input));
+    assert(getShape(_axes).DimensionsCount() == 1);
+    for (int i = 0; i < getShape(_axes).Dims(0); ++i)
+    {
+      keep_dim_shape.SetDim(reinterpret_cast<int32_t *>(_axes->buffer())[i], 1);
+    }
+  }
```
Comment (zetwhite) [inline runtime/onert/backend/train/ops/MeanLayer.cc:63]: I'm not sure about this part. 

I thought axes is just an one integer value (axis = 0, axis = 1 ..).

But It looks like axes is represetend in tensor in onert. :thinking:

Comment (jyoungyun) [reply runtime/onert/backend/train/ops/MeanLayer.cc:63]: Yes, the axes is a tensor in onert. And AFAIK, the `axes` tensor is one dimension tensor. It contains multiple axes to reduce.

This part is not necessary in the formula, but it is necessary to use Broadcast in eigen. If the `keep_dims` is false, this operation reduces the given axis dimension. So if the input shape of X is (2,2) and the axis is 0, the output shape of reduce mean is (2,). In eigen, the (2,) shape can not broadcast to the (2,2). So it changes the (2,) shape to the (1, 2) shape. This part changes the `_back_prop_input` shape to a shape that broadcast is possible.

Comment (zetwhite) [reply runtime/onert/backend/train/ops/MeanLayer.cc:63]: > Yes, the axes is a tensor in onert. And AFAIK, the axes tensor is one dimension tensor. It contains multiple axes to reduce. 

Aha, It contains **multiple** axes to reduce. This point it what I missed. 

Thanks for the kind explanation! I clearly understood :+1:

**Review by zetwhite — APPROVED** [review]
LGTM :+1:

File: runtime/onert/backend/train/ops/MeanLayer.cc [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  else
+  {
+    keep_dim_shape.ReplaceWith(getShape(_back_prop_input));
+  }
```
Comment (ragmani) [inline runtime/onert/backend/train/ops/MeanLayer.cc:67]: (Optional)
You can use `getReducerAxes()` in `OperationUtils.h`.

File: runtime/onert/backend/train/ops/MeanLayer.cc [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+void MeanLayer::forward(bool) { cpu::ops::MeanLayer::run(); }
+
+void MeanLayer::backward()
+{
+  nnfw::cker::Shape keep_dim_shape;
+  if (_keep_dims == false)
+  {
+    keep_dim_shape.ReplaceWith(getShape(_input));
+    assert(getShape(_axes).DimensionsCount() == 1);
```
Comment (ragmani) [inline runtime/onert/backend/train/ops/MeanLayer.cc:58]: We may consider the case that rank is 0.
```suggestion
    assert(_getShape(_axes).DimensionsCount() == 0 || getShape(_axes).DimensionsCount() == 1);
```
or
```suggestion
    assert(_axes->getShape().rank() == 0 || _axes->getShape().rank() == 1);
```

File: runtime/onert/backend/train/ops/MeanLayer.cc (hunk 2 above)
Comment (jyoungyun) [reply runtime/onert/backend/train/ops/MeanLayer.cc]: I applied `getReducerAxes()`. PTAL again.

**Review by zetwhite — APPROVED** [review]

**Review by ragmani — APPROVED** [review]
LGTM

**Comment by jyoungyun** [comment]
@Samsung/one_onert PTAL

**Pull request merged**